core/test/distributed/index_map.cpp

```diff
@@ -0,0 +1,165 @@
+    using part_type =
+        gko::experimental::distributed::Partition<local_index_type,
+                                                  global_index_type>;
+    using map_type =
+        gko::experimental::distributed::index_map<local_index_type,
+                                                  global_index_type>;
+
+    std::shared_ptr<const gko::Executor> exec =
+        gko::ReferenceExecutor::create();
+    std::shared_ptr<part_type> part =
+        part_type::build_from_global_size_uniform(exec, 3, 6);
+};
+
+TYPED_TEST_SUITE(IndexMap, gko::test::LocalGlobalIndexTypes,
+                 PairTypenameNameGenerator);
+
+
+template <typename T>
+void assert_collection_eq(const gko::segmented_array<T>& a,
+                          const gko::segmented_array<T>& b)
+{
+    auto get_flat_array = [](const gko::segmented_array<T>& arr) {
+        return gko::make_const_array_view(arr.get_executor(), arr.get_size(),
+                                          arr.get_const_flat_data());
+    };
+    GKO_ASSERT_ARRAY_EQ(a.get_offsets(), b.get_offsets());
+    GKO_ASSERT_ARRAY_EQ(get_flat_array(a), get_flat_array(b));
+}
+
+
+TYPED_TEST(IndexMap, CanDefaultConstruct)
+{
+    using map_type = typename TestFixture::map_type;
+
+    auto imap = map_type(this->exec);
+
+    ASSERT_EQ(imap.get_local_size(), 0);
+    ASSERT_EQ(imap.get_non_local_size(), 0);
+    ASSERT_EQ(imap.get_remote_target_ids().get_size(), 0);
+    ASSERT_EQ(imap.get_remote_global_idxs().get_size(), 0);
+    ASSERT_EQ(imap.get_remote_local_idxs().get_size(), 0);
+}
+
+
+TYPED_TEST(IndexMap, CanCopyConstruct)
+{
+    using map_type = typename TestFixture::map_type;
+    using global_index_type = typename TestFixture::global_index_type;
+    gko::array<global_index_type> connections(this->exec, {4, 3, 3, 4, 2});
+    auto imap = map_type(this->exec, this->part, 0, connections);
+
+    auto copy = imap;
+
+    GKO_ASSERT_ARRAY_EQ(copy.get_remote_target_ids(),
+                        imap.get_remote_target_ids());
+    assert_collection_eq(copy.get_remote_local_idxs(),
+                         imap.get_remote_local_idxs());
+    assert_collection_eq(copy.get_remote_global_idxs(),
+                         imap.get_remote_global_idxs());
+}
+
+TYPED_TEST(IndexMap, CanMoveConstruct)
+{
+    using map_type = typename TestFixture::map_type;
+    using global_index_type = typename TestFixture::global_index_type;
+    gko::array<global_index_type> connections(this->exec, {4, 3, 3, 4, 2});
+    auto imap = map_type(this->exec, this->part, 0, connections);
+    auto copy = imap;
+    auto imap_remote_global_it =
+        imap.get_remote_global_idxs().get_const_flat_data();
+    auto imap_remote_local_it =
+        imap.get_remote_local_idxs().get_const_flat_data();
+    auto imap_remote_target_it = imap.get_remote_target_ids().get_const_data();
+
+    auto move = std::move(imap);
+
+    GKO_ASSERT_ARRAY_EQ(move.get_remote_target_ids(),
```

Comment:
you could also check the moved-from state here, but we rarely do that, so it's fine if you don't ;)